packages/compiler/src/experimental/mutators.ts

```diff
@@ -1,95 +1,200 @@
+ * @param program - The program in which the `sourceType` occurs.
+ * @param realm - The realm where the `sourceType` will be cloned, if this type is mutated.
+ *
+ * @returns a boolean or {@link MutatorFlow} flags.
+ *
+ * @experimental
+ */
+export type MutatorFilterFn<T extends Type> = (
+  sourceType: T,
+  program: Program,
+  realm: Realm,
+) => boolean | MutatorFlow;
+
+/**
+ * A function that replaces a mutable type with a new type instance.
+ *
+ * Returning `clone` from this function is equivalent to providing a `mutate` function instead of a `replace` function.
+ *
+ * This function runs after the `sourceType` is cloned within the realm.
+ *
+ * @param sourceType - The source type.
+ * @param clone - The clone of the source type to mutate.
+ * @param program - The program in which the `sourceType` occurs.
+ * @param realm - The realm in which the `clone` resides.
+ *
+ * @returns a new type instance to replace the cloned type instance.
+ *
+ * @experimental
+ */
+export type MutatorReplaceFn<T extends Type> = (
+  sourceType: T,
+  clone: T,
+  program: Program,
+  realm: Realm,
+) => Type;
+
+/**
+ * Mutators describe procedures for mutating types in the type graph.
+ *
+ * Each entry in the mutator describes how to mutate a specific type of node.
+ *
+ * See {@link mutateSubgraph}.
+ *
+ * @experimental
+ */
+export type Mutator = {
+  /**
+   * The name of this mutator.
+   */
   name: string;
-  Model?: MutatorRecord<Model>;
-  ModelProperty?: MutatorRecord<ModelProperty>;
-  Scalar?: MutatorRecord<Scalar>;
-  Enum?: MutatorRecord<Enum>;
-  EnumMember?: MutatorRecord<EnumMember>;
-  Union?: MutatorRecord<Union>;
-  UnionVariant?: MutatorRecord<UnionVariant>;
-  Tuple?: MutatorRecord<Tuple>;
-  Operation?: MutatorRecord<Operation>;
-  Interface?: MutatorRecord<Interface>;
-  String?: MutatorRecord<Scalar>;
-  Number?: MutatorRecord<Scalar>;
-  Boolean?: MutatorRecord<Scalar>;
-  ScalarConstructor?: MutatorRecord<ScalarConstructor>;
-  StringTemplate?: MutatorRecord<StringTemplate>;
-  StringTemplateSpan?: MutatorRecord<StringTemplateSpan>;
-}
+} & {
+  /**
+   * Describes how to mutate a type with the given node kind.
+   */
+  [Kind in MutableType["kind"]]?: MutatorRecord<Extract<MutableType, { kind: Kind }>>;
+};
 
 /**
- * @experimental - This is a type that extends Mutator with a Namespace property.
+ * A mutator that can additionally mutate namespaces.
+ *
+ * @experimental
  */
 export type MutatorWithNamespace = Mutator & {
   Namespace: MutatorRecord<Namespace>;
 };
 
-/** @experimental */
+/**
+ * Flow control for mutators.
+ *
+ * When filtering types in a mutator, the filter function may return MutatorFlow flags to control how mutation should
+ * proceed.
+ *
+ * @see {@link MutatorFilterFn}
+ *
+ * @experimental
+ */
 export enum MutatorFlow {
-  MutateAndRecurse = 0,
+  /**
+   * Mutate the type and recur, further mutating the type's children. This is the default behavior.
+   */
+  MutateAndRecur = 0,
```

Comment:
I get that recur is the proper verb form... but recur is also kind of a general term for "might happen again" which is a bit ambiguous and recurse is *I think* commonly used as a verb, but maybe I'm some kind of heathen.

Reply:
Personally I've always heard and said "recur" and never "recurse" (e.g. clojure's `recur` operator) and didn't find a definition for "recurse" in the first dictionary I looked at (Merriam-Webster online), but I see now that many dictionaries do have a definition for it. In the context of formal functional programming, the relationship between a function and its recursive calls was always described to me as its "recurrence relation" and I think that "recursion" and "recurrence" are basically the same thing (after all, the reason it's called "recursion" is because you're defining a thing that recurs within its own definition.

I'm not married to the word "recur", so if you think "recurse" is better I wont object to reverting these back.

Reply:
no strong opinion here so you can leave it if you like!